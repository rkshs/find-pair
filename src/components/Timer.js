import { useEffect, useState, forwardRef, useImperativeHandle } from 'react';

const Timer = forwardRef(({timerStarted, reseter}, ref) => {
    const [time, setTime] = useState({
        second: '00',
        minute: '00',
        counter: 0
    })

    useEffect(() => {
        let intervalId;

        if (timerStarted) {
            intervalId = setInterval(() => {
                const secondCounter = time.counter % 60;
                const minuteCounter = Math.floor(time.counter / 60);

                const computedSecond = String(secondCounter).length === 1 ? `0${secondCounter}` : secondCounter;
                const computedMinute = String(minuteCounter).length === 1 ? `0${minuteCounter}` : minuteCounter;

                setTime(prevState => ({
                    second: computedSecond,
                    minute: computedMinute,
                    counter: prevState.counter + 1
                }))
            }, 1000)
        }

        return () => clearInterval(intervalId);
    }, [timerStarted, time.counter])

    useImperativeHandle(ref, () => ({
        resetTimer() {
            console.log('trigger')
            setTime(() => ({
                second: '00',
                minute: '00',
                counter: 0
            }))
        }
    }))

    return (
        <span className="time">
            <span className="minute">{time.minute}</span>
            <span>:</span>
            <span className="second">{time.second}</span>
        </span>
    )
})

export default Timer;
