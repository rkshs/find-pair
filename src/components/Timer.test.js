import { cleanup } from '@testing-library/react'
import React, { useRef } from 'react';
import renderer from 'react-test-renderer'
import '@testing-library/jest-dom'
import Timer from './Timer';

jest.mock('react', () => {
    return {
        ...jest.requireActual('react'),
        useRef: jest.fn(),
    };
});

const useMockRef = jest.mocked(useRef);

afterEach(cleanup);

describe('Timer component', () => {
    it('snapshot renders correctly, truthy values', () => {
        const cardView = renderer
            .create(<Timer
                timerStarted={false}
                ref={useMockRef}
            />).toJSON()
        expect(cardView).toMatchSnapshot()
    });
    it('snapshot renders correctly, falsy values', () => {
        const cardView = renderer
            .create(<Timer
                timerStarted={''}
                ref={null}
            />).toJSON()
        expect(cardView).toMatchSnapshot()
    });
});

