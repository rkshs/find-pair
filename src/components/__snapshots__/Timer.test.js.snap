// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Timer component snapshot renders correctly, falsy values 1`] = `
<span
  className="time"
>
  <span
    className="minute"
  >
    00
  </span>
  <span>
    :
  </span>
  <span
    className="second"
  >
    00
  </span>
</span>
`;

exports[`Timer component snapshot renders correctly, truthy values 1`] = `
<span
  className="time"
>
  <span
    className="minute"
  >
    00
  </span>
  <span>
    :
  </span>
  <span
    className="second"
  >
    00
  </span>
</span>
`;
